server.js

```diff
@@ -5,15 +5,79 @@ const request = require('request');
 const User = require('./models/User')
 const Image = require('./models/Image')
 
+const bcrypt = require('bcrypt');
 const app = express();
 app.use(cors())
 app.use(express.json());
+app.use((req, res, next) => {
+  console.log(req.body);
+  next();
+});
 
 const sequelize = new Sequelize('hamishgpt', 'user', 'root', {
   host: 'db',
   dialect: 'mysql'
 });
 
+app.get('/ping', function (request, response) {
+  response.send("pong");
+});
+
+app.post('/registration', async (request, response) => {
+  const { username, email, password } = request.body;
+  if (username === null || username.length < 1) {
+    return response.status(400).send('Username is invalid');
+  }
+  if (email === null || email.length < 1) {
+    return response.status(400).send('Email is invalid');
+  }
+  var validRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
+  if (!email.match(validRegex)) {
+    return response.status(400).send('Invalid email address');
+  }
+  if (password === null || password.length < 8 || password.length > 20) {
+    return response.status(400).send('Password is invalid');
+  }
+  try {
+    const user = await User.create({
+      username: request.body.username,
+      email: request.body.email,
+      password: request.body.password
+    })
+    if (user) {
+      response.send(user);
+    } else {
+      return response.status(400).send('Error in insert new record');
+    }
+  } catch (error) {
+    console.error(error);
+    response.status(400).send(error.errors[0].message);
+  }
+});
+
+app.post('/login', async (req, res) => {
+  console.log('Login endpoint hit. Request body:', req.body)
+  const { username, password } = req.body;
+
+  const user = await User.findOne({ where: { username: username } });
+  if (!user) {
+    return res.status(400).json({ error: 'User not found' });
+  }
+
+  const validPassword = await bcrypt.compare(password, user.password);
+  if (!validPassword) {
+    return res.status(400).json({ error: 'Invalid password' });
+  }
+  console.log('Login successful:', user);
+  res.json({

```

Comment:
Just as a POC and show how the login is working, you can return those sensitive data in clear. But as soon as you understand what you should return you will have to replace this.

Reply:
I've added a token authentication library and removed the sensitive data in replacement of the token instead. Is this a good way?